### DISCUSSION TITLE
running keycloak with service account

### DISCUSSION THREAD

**Discussion by da-development-id** [keycloak/keycloak#30150] [Q&A] [answered]
Trying to run keycloak with a service account:

`sudo -u keycloak /opt/keycloak/bin/kc.sh start --optimized`

but I'm getting the error:

`Exception in thread "main" java.lang.NoClassDefFoundError: io/smallrye/common/io/jar/JarFiles
    at io.quarkus.bootstrap.runner.JarResource.ensureJarFileIsOpen(JarResource.java:170)
    at io.quarkus.bootstrap.runner.JarResource.readLockAcquireAndGetJarReference(JarResource.java:153)
    at io.quarkus.bootstrap.runner.JarResource.getResourceData(JarResource.java:72)
    at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:99)
    at io.quarkus.bootstrap.runner.RunnerClassLoader.loadClass(RunnerClassLoader.java:65)
    at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:60)
    at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.lang.ClassNotFoundException: io.smallrye.comon.io.jar.JarFiles
    at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
    at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
    at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
    ... 7 more`

Environment:
RHEL 8.9
Keycloak 22.0.1
Keycloak directory/files owned by keycloak:keycloak
Java is RHEL java-17-openjdk

Running as root seems to be OK.  Disabling SELinux does not resolve issue as well.

**Comment by codespearhead** [comment]
Make sure the service account has the necessary permissions and try again:

```bash
sudo chown -R keycloak:keycloak /opt/keycloak
```

**Reply by da-development-id** [reply]
yes, they are already owned by keycloak:keycloak

**Reply by codespearhead** [reply]
Are you using the community build or the Red Hat build of Keycloak?

If it's the former, is the issue reproducible in the latest major version of Keycloak (24.0.5)?

**Reply by da-development-id** [reply]
It is the community version, I will need to try 24.0.5

**Reply by da-development-id** [reply]
are there specific permisions needed by the user, group, and/or other?

**Reply by codespearhead** [reply]
No, there isn't.

Try this:

```bash
sudo chmod -R u+rX /opt/keycloak
sudo -u keycloak /opt/keycloak/bin/kc.sh start --optimized
```

And if that doesn't work, try this:

```bash
sudo -E -u keycloak /opt/keycloak/bin/kc.sh start --optimized
```

**Reply by da-development-id** [reply]
from the audit log:
node=myserver type=SYSCALL msg=audit(1717780989.610:5373):  arch=c000003e syscall=257 success=no exit=-1 a0=ffffff9c a1=7fd53414c610 a2=0 a3=0 items=1 ppid=7691 pid=7694 auid=1178401137 uid=983 gid=1001 euid=983 suid=983 fsuid=983 egid=1001 sgid=1001 fsgid=1001 tty=pts1 ses=5 comm="java" exe="/usr/lib/jvm/java-17-openjdk-17.0.11.0.9-2.el8.x86_64/bin/java" subj=staff_u:sysadm_r:sysadm_t:s0-s0:c0.c1023 key="access"  ARCH=x86_64 SYSCALL=openat AUID="myuser" UID="keycloak" GID="keycloak" EUID="keycloak" SUID="keycloak" FSUID="keycloak" EGID="keycloak" SGID="keycloak" FSGID="keycloak"

**Reply by da-development-id** [reply]
turns out to be fapolicyd that is blocking access

**Reply by codespearhead** [reply]
Is this solved then? If so, can you mark one of those comments [as the answer to the discussion](https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#marking-a-comment-as-an-answer) so that it doesn't show up [in the discussion tab](https://github.com/keycloak/keycloak/discussions) as unanswered?

**Comment by da-development-id** [comment, marked as answer]
Looks like I will have to do something like the following, to add to the trusted files:

```
# sudo fapolicyd-cli --file add /opt/keycloak --trust-file keycloak
# sudo fapolicyd-cli --update
```

or maybe create a policy that allows files marked as 'application/java-archive', which seems to be denied by default. some of the jars are marked as 'application/zip' and those seems to load without issue.  I think I prefer the former option above since it is more restrictive and just update the trusted files when updates are deployed.
